src/platform/riot/uart_channel.c

```diff
@@ -0,0 +1,231 @@
+
+  self->receive_buffer[receive_buffer_index] = received_byte;
+
+  if (self->receive_buffer_index >= minimum_message_size) {
+    if (self->super.super.mode == NETWORK_CHANNEL_MODE_ASYNC) {
+      cond_signal(&((UARTAsyncChannel *)self)->receive_cv);
+    }
+  }
+}
+
+void UARTPollChannel_poll(NetworkChannel *untyped_self) {
+  UARTPollChannel *self = (UARTPollChannel *)untyped_self;
+  const uint32_t minimum_message_size = 12;
+
+  while (self->receive_buffer_index > minimum_message_size) {
+    int bytes_left = deserialize_from_protobuf(&self->output, self->receive_buffer, self->receive_buffer_index);
+    UART_CHANNEL_DEBUG("Bytes Left after attempted to deserialize %d", bytes_left);
+
+    if (bytes_left >= 0) {
+
+      self->env->enter_critical_section(self->env);
+      int receive_buffer_index = self->receive_buffer_index;
+      self->receive_buffer_index = bytes_left;
+      self->env->leave_critical_section(self->env);
+
+      memcpy(self->receive_buffer, self->receive_buffer + (receive_buffer_index - bytes_left), bytes_left);
+
+      if (self->receive_callback != NULL) {
+        UART_CHANNEL_DEBUG("calling user callback!");
+        self->receive_callback(self->federated_connection, &self->output);
+      }
+    } else {
+      break;
+    }
+  }
+}
+
+void *_UARTAsyncChannel_decode_loop(void *arg) {
+  UARTAsyncChannel *self = (UARTAsyncChannel *)arg;
+  mutex_lock(&self->receive_lock);
+
+  UART_CHANNEL_DEBUG("Entering decoding loop");
+  while (true) {
+    cond_wait(&self->receive_cv, &self->receive_lock);
+
+    UARTPollChannel_poll((NetworkChannel *)&self->super);
+  }
+
+  return NULL;
+}
+
+void UARTPollChannel_ctor(UARTPollChannel *self, Environment *env, uint32_t uart_device, uint32_t baud,
+                          uint8_t data_bits, uint8_t parity, uint8_t stop_bits) {
+  assert(self != NULL);
+  assert(env != NULL);
+
+  self->uart_dev = UART_DEV(uart_device);
+
+  int result = uart_init(self->uart_dev, baud, _UARTPollChannel_interrupt_callback, self);
+
+  if (result == -ENODEV) {
+    UART_CHANNEL_ERR("Invalid UART device!");
+    throw("The user specified an invalid UART device!");
+  } else if (result == ENOTSUP) {
+    UART_CHANNEL_ERR("Given configuration to uart is not supported!");
+    throw("The given combination of parameters for creating a uart device is unsupported!");
+  } else if (result < 0) {
+    UART_CHANNEL_ERR("UART Init error occurred");
+    throw("Unknown UART RIOT Error!");
+  }
+
+  uart_data_bits_t riot_data_bits;
+
+  switch (data_bits) {
+  case 5:
+    riot_data_bits = UART_DATA_BITS_5;
+    break;
+  case 6:
+    riot_data_bits = UART_DATA_BITS_6;
+    break;
+  case 7:
+    riot_data_bits = UART_DATA_BITS_7;
+    break;
+  case 8:
+    riot_data_bits = UART_DATA_BITS_8;
+    break;
+  default:
+    UART_CHANNEL_ERR("Invalid data_bits given to UARTPollChannel %u", data_bits);
+    throw("Invalid data_bits given!");
+  };
+
+  uart_parity_t riot_parity;
+
+  switch (parity) {
```

Comment:
Not to be a nitpicker, but wouldn't it be nicer to define our own enum for this. And while we are at it, we could define it for data_bits and stop_bits also?

Reply:
I agree I went for integers because if we define an enum that this enum is not platform specific and should not be definded inside platform specific code.

Reply:
Where should this enum go? 

- platform
   - riot
      - uart_channel.h
- network_channel
    - tcp_ip_channel.h
   - uart_channel.h <-- inside here